src/dist/temp.rs

```diff
@@ -158,7 +158,7 @@ impl Cfg {
     }
 
     pub(crate) fn clean(&self) {
-        utils::delete_dir_contents(&self.root_directory);
+        utils::delete_dir_contents_following_links(&self.root_directory);
```

Comment:
this should never be a symlink, it is created on line 105. Its probably ok to permit following links in order to get this regression fixed, but longer term we should only be deleting dirs here, since getting a symlink is impossible without someone altering state

Reply:
@rbtcollins Hmmm https://github.com/rust-lang/rustup/issues/3344 explicitly mentioned the case where `tmp` is a link, so I'm not quite sure about that.

Reply:
You added a regression test for that, so empirical validatation of whether its the one or not should be quite easy. I don't think it is as this is the tmp root for one dist operation, not the dir that temp operations are rooted in.

Reply:
> You added a regression test for that, so empirical validatation of whether its the one or not should be quite easy.

@rbtcollins Replacing this line with the old implementation gives:

```console
out.stderr (18 lines):
    info: syncing channel updates for 'nightly-aarch64-apple-darwin'
    info: latest update on 2015-01-02, rust version 1.3.0 (hash-nightly-2)
    info: downloading component 'cargo'
    info: downloading component 'rust-docs'
    info: downloading component 'rust-std'
    info: downloading component 'rustc'
    info: removing previous version of component 'cargo'
    info: removing previous version of component 'rust-docs'
    info: removing previous version of component 'rust-std'
    info: removing previous version of component 'rustc'
    info: installing component 'cargo'
    info: installing component 'rust-docs'
    info: installing component 'rust-std'
    info: installing component 'rustc'
    info: cleaning up downloads & tmp directories
    thread 'main' panicked at src/utils/utils.rs:616:13:
    Unable to clean up <snip>/rustupRd3iIp/tmp: Os { code: 62, kind: FilesystemLoop, message: "Too many levels of symbolic links" }
```

So I guess both pieces are indeed required for the fix to work.